src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/AnalyzerAssemblyService.cs

```diff
@@ -0,0 +1,204 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Reflection;
+using System.Reflection.Metadata;
+using System.Reflection.PortableExecutable;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Tree.Dependencies
+{
+    [Export(typeof(IAnalyzerAssemblyService))]
+    [AppliesTo(ProjectCapability.DependenciesTree)]
+    internal class AnalyzerAssemblyService : IAnalyzerAssemblyService
+    {
+        private object _guard = new object();
+        private Dictionary<string, bool> _map = new Dictionary<string, bool>(StringComparer.OrdinalIgnoreCase);
+
+        public bool ContainsDiagnosticAnalyzers(string fullPathToAssembly)
+        {
+            lock (_guard)
+            {
+                if (!_map.TryGetValue(fullPathToAssembly, out bool containsDiagnosticAnalyzers))
```

Comment:
What are the threading rules here? will be called from multiple threads?

Reply:
I assume it will, yes.

Reply:
If that is the case, is there a way we could asynchronously wait for the lock instead of blocking all calling threads?

Reply:
I'll check. I can certainly use a `SemaphoreSlim` as an asynchronous lock, but I'm not sure if the code path calling into the cache can be made asynchronous.

Reply:
Can we make sure that read is can be done without waiting?

Reply:
I've updated the code so the lock is only taken long enough to update the dictionary.